lib/ldclient-rb/ldclient.rb

```diff
@@ -131,6 +132,23 @@ def initialize(sdk_key, config = Config.default, wait_for_sec = 5)
       end
     end
 
+    #
+    # Add a hook to the client. In order to register a hook before the client starts, please use the `hooks` property of
+    # {#LDConfig}.
+    #
+    # Hooks provide entrypoints which allow for observation of SDK functions.
+    #
+    # @param hook [Interfaces::Hooks::Hook]
+    #
+    def add_hook(hook)
+      unless hook.is_a?(Interfaces::Hooks::Hook)
+        @config.logger.error { "[LDClient] Attempted to add a hook that does not include the LaunchDarkly::Intefaces::Hooks:Hook mixin. Ignoring." }
+        return
+      end
+
+      @hooks.push(hook)
```

Comment:
I don't remember, does ruby have threading concerns?

Reply:
Like your node implementation, I make a copy of the hooks before I begin processing, so I don't think there is a thread issue here.

Reply:
Ok. I wasn't sure if you actually had to lock the collection. In node there aren't threads, so it doesn't have to worry about any memory synchronization stuff.

Reply:
Actually in thinking about it more, I do need to do some locking. This would work fine in MRI, but jRuby would probably have some sync issues since it doesn't use a GIL.

Updated to use the concurrent package we have used elsewhere.